File: go/0003.go
/* file: 0003.go
 * by  : agaric
 * copy: public domain
 * desc: project euler #3 - "largest prime factor"
 * lang: go
 */

package main

import ("fmt"; "math")

func sieve(n int) []int {
	a := []int{}
	b := make([]bool, n)
	for i, _ := range b {
		b[i] = true
	}
	b[0] = false
	if n >= 1 { b[1] = false }
	for i, x := range b {
		if x {
			m := i * i;
			for m < n {
				b[m] = false
				m += i;
			}
		}
	}
	for i, x := range b {
		if x { a = append(a, i) }
	}
	return a
}

func main() {
	n := 600851475143
	p := sieve(int(math.Sqrt(float64(n))))
	max := 1
	for _, x := range p {
		if n % x == 0 && x > max {
			max = x
		}
	}
	fmt.Println(max)
}

File: rust/0003.rs
/* file: 0003.rs
 * by  : agaric
 * desc: project euler #3 - "largest prime factor"
 */

fn sieve(n: usize) -> Vec<usize> {
    let mut v = vec![true; n + 1];
    v[0] = false;
    if n >= 1 { v[1] = false; }

    for i in 2..(n + 1) {
        if v[i] {
            let mut m = i * i;
            while m <= n {
                v[m] = false;
                m += i;
            }
        }
    }

    v.iter().enumerate()
        .filter_map(|(p, &is_p)| if is_p {Some(p)} else {None})
        .collect()
}

fn main() {
    let x = 600851475143usize;
    let primes = sieve((x as f64).sqrt() as usize);
    let mut max = 1;
    for p in primes {
        if x % p == 0 && p > max {
            max = p;
        }
    }
    println!("{}", max);
}
